### PR TITLE
llvm upgrade: apply applicable old patches

### PR DESCRIPTION
**Pull request by vtjnash**
~~A couple have not yet been updated, but this is a start towards updating the patch list~~ Only one is missing now for LLVM 8. LLVM 7 is complete.

~~Also, this currently fails an offsetarray test.~~ tracked this down to bad TBAA information and fixed it

Needs nanosoldier, and Keno suggested it should have a PkgEval run too.

### CODE REVIEW

File: deps/llvm.mk [hunk 1]
```diff
@@ -431,7 +431,34 @@ endif
...
+$(eval $(call LLVM_PATCH,llvm7-D51842-win64-byval-cc)) # remove for 8.0
+$(eval $(call LLVM_PATCH,llvm-D57118-powerpc))
+endif # LLVM_VER 7.0
+
+ifeq ($(LLVM_VER_SHORT),8.0)
+$(eval $(call LLVM_PATCH,llvm-D27629-AArch64-large_model_6.0.1))
+$(eval $(call LLVM_PATCH,llvm8-D34078-vectorize-fdiv))
+$(eval $(call LLVM_PATCH,llvm-6.0-NVPTX-addrspaces)) # NVPTX -- warning: this fails check-llvm-codegen-nvptx
+$(eval $(call LLVM_PATCH,llvm-7.0-D44650)) # mingw32 build fix
+#$(eval $(call LLVM_PATCH,llvm-D46460))
```
Comment (Keno) [inline deps/llvm.mk]: This one turns out to not be required anymore because the issue got fixed upstream

File: deps/llvm.mk [hunk 2]
```diff
@@ -431,7 +431,34 @@ endif
...
+$(eval $(call LLVM_PATCH,llvm-6.0-DISABLE_ABI_CHECKS))
+$(eval $(call LLVM_PATCH,llvm7-D50010-VNCoercion-ni))
+#$(eval $(call LLVM_PATCH,llvm-D50167-scev-umin))
+$(eval $(call LLVM_PATCH,llvm7-windows-race))
+$(eval $(call LLVM_PATCH,llvm-D57118-powerpc))
```
Comment (Keno) [inline deps/llvm.mk]:
```suggestion
$(eval $(call LLVM_PATCH,llvm-D57118-powerpc)) # remove for 9.0
```

File: deps/llvm.mk (hunk 1 above)
Comment (Keno) [reply deps/llvm.mk]:
```suggestion
```

File: deps/llvm.mk [hunk 3]
```diff
@@ -431,7 +431,34 @@ endif
...
+
+ifeq ($(LLVM_VER_SHORT),7.0)
+$(eval $(call LLVM_PATCH,llvm-D27629-AArch64-large_model_6.0.1))
+$(eval $(call LLVM_PATCH,llvm-D34078-vectorize-fdiv))
+$(eval $(call LLVM_PATCH,llvm-6.0-NVPTX-addrspaces)) # NVPTX -- warning: this fails check-llvm-codegen-nvptx
+$(eval $(call LLVM_PATCH,llvm-7.0-D44650)) # mingw32 build fix
+$(eval $(call LLVM_PATCH,llvm-D46460))
+$(eval $(call LLVM_PATCH,llvm-6.0-DISABLE_ABI_CHECKS))
+$(eval $(call LLVM_PATCH,llvm7-D50010-VNCoercion-ni))
+#$(eval $(call LLVM_PATCH,llvm-D50167-scev-umin))
```
Comment (vchuravy) [inline deps/llvm.mk]: Keno backported that in #28809

**Review by vchuravy — COMMENTED** [review]
Needs the version bump for windows

**Comment by vtjnash** [comment]
Almost working now (doesn't need to block merging though).
- One bug in a new X86 pass: https://bugs.llvm.org/show_bug.cgi?id=41629 (just disabled it for now)
- And one bug that I've isolated down to coming out of LICM (it moves a store after its use): https://gist.github.com/vtjnash/8a073c01e0a6e6b29789bb3f7080e72b, but haven't yet been able to trigger this outside of Julia

**Comment by vtjnash** [comment]
Ah, it wasn't LICM's fault (sorry LICM), someone (emit_select) was creating bad TBAA data for it. We should probably backport that commit fix.

**Comment by vchuravy** [comment]
Probably separate f826bba out into it's own PR? It currently fails tests:

```
      From worker 6:	signal (6): Aborted
      From worker 6:	in expression starting at /buildworker/worker/tester_linux32/build/share/julia/test/bitarray.jl:1097
      From worker 6:	gsignal at /lib/i386-linux-gnu/libc.so.6 (unknown line)
      From worker 6:	abort at /lib/i386-linux-gnu/libc.so.6 (unknown line)
      From worker 6:	unknown function (ip: 0xf72780f5)
      From worker 6:	__assert_fail at /lib/i386-linux-gnu/libc.so.6 (unknown line)
      From worker 6:	mark_julia_slot at /buildworker/worker/package_linux32/build/src/codegen.cpp:624
      From worker 6:	mark_julia_slot at /buildworker/worker/package_linux32/build/src/intrinsics.cpp:854 [inlined]
      From worker 6:	emit_ifelse at /buildworker/worker/package_linux32/build/src/intrinsics.cpp:851 [inlined]
```

Would be good to have the backport for the scev patch for LLVM 8, otherwise LGTM.

cc: @banex19 since this might impact you.

**Comment by Keno** [comment]
https://gist.github.com/Keno/0107cea84de827845d204fc398d877ff for a llvm 8.0 version of the scev patch. It's a backport of the current state of the review rather than a forward port of the current patch, FYI.

File: src/llvm-api.cpp [hunk 4]
```diff
@@ -220,11 +220,20 @@ extern "C" JL_DLLEXPORT LLVMContextRef LLVMExtraGetValueContext(LLVMValueRef V)
...
+
+#if JL_LLVM_VERSION < 80000
 extern ModulePass *createNVVMReflectPass();
 extern "C" JL_DLLEXPORT void LLVMExtraAddNVVMReflectPass(LLVMPassManagerRef PM)
 {
     unwrap(PM)->add(createNVVMReflectPass());
 }
+#else
+FunctionPass *createNVVMReflectPass(unsigned int SmVersion);
+extern "C" JL_DLLEXPORT void LLVMExtraAddNVVMReflectFunctionPass(LLVMPassManagerRef PM, unsigned int SmVersion)
```
Comment (vchuravy) [inline src/llvm-api.cpp]: cc: @maleadt

**Review by vchuravy — APPROVED** [review]
LGTM

**Pull request merged**